### PR TITLE
test/testutil: Simplify API

### PR DESCRIPTION
**Pull request by ccollins476ad**
### Notes
* This a compatibility breaking change to the testutil API.  Initially, I was just going to deprecate the old API, but keep it in the codebase for one more release.  Frankly, I think this is way more effort than it is worth.  I doubt any packages outside of mynewt-core are using testutil at all.  If this is overreaching, I can deprecate the old API properly.

* If you look at all the commits at once, this PR will seem insane.  I would focus on the `test/testutil` commit (`2766a47`).  All the other changes just convert clients to the new API and cleanup existing unit tests.

### Description
This a compatibility breaking change to the testutil API.

Simplify the `testutil` API as follows:

1. Add the `TEST_CASE_SELF()` macro.  This macro creates a test case suitable for a self test (runnable via `newt test`).  These test cases call `sysinit()` automatically when they start.

2. Removing several callbacks from the testutil configuration.

These changes are described in more detail in this email: https://lists.apache.org/thread.html/ec5d5ac556b1b955a60efac05b673e729fbf89bdb4675a1e34a1063a@%3Cdev.mynewt.apache.org%3E

### CODE REVIEW

**Comment by mkiiskila** [comment]
Do all unittest packages need changes due to this?
E.g. Kevin is a user of testutil, afaik all his repositories come with unit tests.
https://github.com/microbuilder/mb_ht16k33
https://github.com/microbuilder/mb_interpolate
https://github.com/microbuilder/mb_sformat

**Comment by ccollins476ad** [comment]
Thanks, @mkiiskila.  No, not all unittest packages need to change.  The `TEST_CASE` macro should continue to work as before, so those `microbuilder` unit tests should continue to work (and will always work).

HOWEVER, this PR does make some compatibility breaking changes to the testutil configuration API.  I don't believe those microbuilder tests use any parts of the discarded API, but other people's tests might.  After thinking about it some more, I don't think it is acceptable to break this API without any heads up.  I will update this PR to preserve backwards compatibility, but will mark deprecated functions with the `deprecated` gcc attribute.  The fact that you were able to find some test code so quickly tells me that there must be more testutil users than I guessed!

So I will mark this as do-not-merge until I restore backwards compatibility.  Thanks for the heads up!

**Comment by ccollins476ad** [comment]
I have re-added the parts of the API that I removed earlier, but I have marked them "deprecated".

**Comment by utzig** [comment]
@ccollins476ad The `net/oic` and `net/ip/mn_socket` tests are known to fail on the CI and are "blacklisted": https://github.com/JuulLabs-OSS/mynewt-travis-ci/blob/master/test_all.sh#L26. You should probably also add these new "selftest" style tests to this list (maybe make a decent blacklist!)

File: encoding/json/selftest/src/test_json_utils.c [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+char *output1 ="{\"KeyBoolArr\": [true, false], \"KeyUintArr\": [0, 65535, 4294967295, 8589934590, 3451257]}";
+char *outputboolspace = "{\"KeyBoolArr\": [    true    ,    false,true         ]}";
+char *outputboolempty = "{\"KeyBoolArr\": , \"KeyBoolArr\": [  ]}";
+
+char *bigbuf;
+int buf_index;
+
+int
+test_write(void *buf, char* data, int len) {
```
[the remark this answers is not in the document]
Comment (ccollins476ad) [reply encoding/json/selftest/src/test_json_utils.c]: Thanks @mkiiskila.  That JSON test code is indeed a bit weird.  I have fixed the style issues that I saw.

**Comment by ccollins476ad** [comment]
Thanks, @utzig - I hadn't noticed that.  I have submitted a PR to the CI repo with that change: https://github.com/JuulLabs-OSS/mynewt-travis-ci/pull/39

**Pull request merged**